medallion/test/config/test_config_loading.py

```diff
@@ -0,0 +1,250 @@
+import json
+import json.decoder
+from unittest import mock
+
+import pytest
+import pytest_subtests  # noqa: F401
+
+import medallion.config as m_cfg
+
+
+def test_load_config_defaults():
+    """
+    Ensure that `conf_file` and `conf_dir` arguments have expected defaults.
+
+    We do this since we don't want to attempt to create those files for parsing
+    in any of our actual functional tests below.
+    """
+    assert m_cfg.load_config.__defaults__ == (
+        m_cfg.DEFAULT_CONFFILE, m_cfg.DEFAULT_CONFDIR,
+    )
+
+
+def test_load_config_default_file_missing(subtests, tmp_path):
+    """
+    Confirm that nothing bad happens if the default config file is missing.
+    """
+    with mock.patch(
+        "medallion.config.DEFAULT_CONFFILE", str(tmp_path / "medallion.conf"),
+    ) as mock_conf_file_default:
+        with subtests.test(msg="conf_dir omitted"):
+            config = m_cfg.load_config(conf_file=mock_conf_file_default)
+            assert config == {}
+        with subtests.test(msg="conf_dir is None"):
+            config = m_cfg.load_config(
+                conf_file=mock_conf_file_default, conf_dir=None,
+            )
+            assert config == {}
+
+
+def test_load_config_non_default_file_missing(subtests, tmp_path):
+    """
+    Confirm an exception is raised if a non-default config file is missing.
+    """
+    conf_file = tmp_path / "missing.conf"
+    with pytest.raises(FileNotFoundError, match=repr(str(conf_file))):
```

Comment:
Because of Windows paths, this needs an `re.escape` for the match, like so:

`match=re.escape(repr(str(conf_file))))`

Reply:
Ah good point! Did you run the tests on a Windows machine? I don't have any dev machines which I could use easily at the moment so if you are able to flush out any other issues which I would have missed because of that it would be helpful :) I'll amend this and the one you tagged below when I get some time tomorrow. Thanks

Reply:
Yeah, made sure to run this with both Unix and Windows. This seems like the only issue though (there may be some issues with environment variables, but I still have to test that).

Reply:
Made this change and just pushed an amended diff. Let me know if that works on your end. I tried to work out if it would play nicely using a variant of this test and passing in a `WindowsPurePath` but the new code creates a `Path` from whatever we pass in so there's some double escaping happening which makes it annoying for me to validate that the test might function as expected on Windows.